doc: add post-release step (#2170)
operator-framework/operator-sdk: doc/dev/release.md

```Markdown
[Version as title, ex. v1.2.3]

**NOTE** See how to upgrade your project to this version by checking the [Version Upgrade Guide](https://github.com/operator-framework/operator-sdk/blob/master/doc/migration/version-upgrade-guide.md)

> 

Author: @joelanford f you folks accepted then, I am committed to go there and update all.

Author: Add it in the post-release actions.

If the release is for a patch version (e.g. `v1.3.1`), an identical PR should be created, targetting the  `v1.3.x` branch. Once this PR passes CI and is merged, `v1.3.x` can be unfrozen.

2. A new section for upgrading operator projects to v1.3.0 should be added to the [version upgrade guide][version-upgrade-guide]. This section should contain steps directing users to upgrade dependency versions in `go.mod`, update function signatures, etc.

3. Then, update the GitHub Release Notes as:

```

> **Note:** See how to upgrade your project to the version v1.3.0 by checking the [Version Upgrade Guide](../migration/version-upgrade-guide.md#v1.3.0)

Author: ```suggestion
**Note:** See how to upgrade your project to the version v1.12.0 by checking the [Version Upgrade Guide](../master/doc/migration/version-upgrade-guide.md#v012x)
```

brew bump-formula-pr --strict --url=$OPERATORSDKURL --sha256=$OPERATORSUM operator-sdk
```

> **Note**  If there were any changes made to the CLI commands, make sure to look at the existing tests, in case they need updating.

Reviewer: ```suggestion
**Note:**  If there were any changes made to the CLI commands, make sure to look at the existing tests, in case they need updating.
```

You've now fully released a new version of the Operator SDK. Good work! Make sure to follow the post-release steps below.

> ### 6. Update the Upgrade version guide document

Reviewer: This step needs to be done _before_ we create a release tag, then the `See how to upgrade your project...` release note needs to be added when we're creating the release notes initially (since this upgrade guide section is tagged in the release).

> ### 6. Update the Upgrade version guide document

Reviewer: ```suggestion
### 6. Update the version upgrade guide
```

> 1. The [Upgrade version guide][version-upgrade-guide] should be updated with the steps required for the users are able to migrate their project.

Reviewer: ```suggestion
1. A new section for upgrading operator projects to v1.3.0 should be added to the [version upgrade guide][version-upgrade-guide] _only if_ there are breaking scaffold or API changes between v1.2.0 and v1.3.0. This section should contain steps directing users to upgrade dependency versions in `go.mod`, update function signatures, etc.
```

Reviewer: This section should be [discussed here](https://github.com/operator-framework/operator-sdk/blob/3b2f684/doc/dev/release.md#release-notes).

Author: Hi @estroz, 

Regards:
> 1. A new section for upgrading operator projects to v1.3.0 should be added to the [version upgrade guide][version-upgrade-guide] _only if_ there are breaking scaffold or API changes between v1.2.0 and v1.3.0. This section should contain steps directing users to upgrade dependency versions in `go.mod`, update function signatures, etc.

The user always will need to perform changes in order to get the upper version. (at least the go.mod and docker file will be needed to updated) Then, it means that always will be required migration steps for a new version. So, added all less the _only_ statement.

> 1. Then, update the GitHub Release Notes as:

Reviewer: This section should be merged with [this list](https://github.com/operator-framework/operator-sdk/blob/3b2f684/doc/dev/release.md#3-create-a-pr-for-post-release-version-and-changelogmd-updates).

Author: We have not the migration steps and the tag in step 3 as suggested. So, it cannot be done before.

Reviewer: The tag will be available on release, and after adding the v1.3.0 upgrade guide section we will be able to properly reference the upgrade guide in the release notes.

Author: OK. I got what you mean. But then, the link will be broken until we add the steps but it is fine. I will move it to there.

Author: Done @estroz 👍

Reviewer: The step adding the linked section to the upgrade guide should be completed before this one during a release, since we need to create an update guide before we tag a release.

Author: We cannot do the migration steps before the release. Note that for we be able to ensure and check them we need the tag released. So, it actually should be made or reviewed by performing them when we are upgrading the samples.

### 1. Update the Upgrade version guide document and Go Memcached Sample

> 1. The [Upgrade version guide][version-upgrade-guide] should be updated with the steps required for the users are able to migrate their project. Then, in order to know the steps and ensure that they will work upgrade the [Go Memcached Sample][go-memcached-sample].

Reviewer: This section is still in the post-release steps. All the information needed to create this upgrade guide section is available when we make a release, so I don't see why this shouldn't be a release step. Move this to its own section in the release steps.

Reviewer 2: Also still need to separate out the version upgrade guide updates from the samples updates. IIRC, we can't do the samples updates until after the release, because the samples require the new tag and base images.

Author: Done

```

## Post-release steps

For the following step should create an JIRA ticket assigned to either yourself or another team member.

> ### 1. Update the Samples and create release tags and notes for the operator-sdk-samples repo 

Reviewer: ```suggestion
### 1. Update and tag the samples repo
```

brew bump-formula-pr --strict --url=$OPERATORSDKURL --sha256=$OPERATORSUM operator-sdk
```

**NOTE**  If there were any changes made to the CLI commands, make sure to look at the existing tests, in case they need updating.

You've now fully released a new version of the Operator SDK. Good work! Make sure to follow the post-release steps below.

### 6. Create JIRA tasks for the post-release actions

After we do the release the following post-release actions are required to attend the users needs. So, create a JIRA for the post release actions  and ping them in the channel in order to allow other sign themselves to help with.

**NOTE** You are free to working on in the post-release tasks if you wish as well. 

> 

Author: However, the goal of this change was to add the step to create the JIRAs. So, if we remove it has no need to do this PR at all. We can just close it. wdyt?

> 

Reviewer: I would at least move this section to the post-release steps, and reword it so that it is a section description:

## Post-release steps

Each of the following steps should have an associated JIRA ticket assigned to either yourself or another team member.

### 1. Update the operator-sdk-samples repo 

...

> 

Author: Done.

## Post-release steps

> ### 1. Update the Upgrade version guide document and Go Memcached Sample

Reviewer 2: ```suggestion
### 1. Update the version upgrade guide
```

> 1. The [Upgrade version guide][version-upgrade-guide] should be updated with the steps required for the users are able to migrate their project. Then, in order to know the steps and ensure that they will work upgrade the [Go Memcached Sample][go-memcached-sample].

Reviewer 2: ```suggestion
1. The [version upgrade guide][version-upgrade-guide] should be updated with the steps required for the users are able to migrate their project. Then, in order to know the steps and ensure that they will work upgrade the [Go Memcached Sample][go-memcached-sample].
```

1. Then, update the GitHub Release Notes as:

```

> **NOTE** See how to upgrade your project to the version v1.3.0 by checking the [Version Upgrade Guide](../migration/version-upgrade-guide.md#v1.3.0)

Reviewer 2: Both sections talk about updating the samples. Let's remove the duplication.

Author: Solved.

> **NOTE** See how to upgrade your project to the version v1.3.0 by checking the [Version Upgrade Guide](../migration/version-upgrade-guide.md#v1.3.0)

Reviewer 2: Let's combine the release notes updates into one step.

Author: I am not sure if I got this one. It is in one step. See the line 326.

Reviewer: Shouldn't creating an upgrade guide section be part of the release and not post-release?

Author: I think it is now as you are looking for @estroz. Could you please check it after the changes.

> ```

Author: @joelanford 
I will wait for we merge this one, then I will do these updates for all versions.

> ### 2. Update the Samples and create release tags and notes for the operator-sdk-samples repo 

Reviewer 2: ```suggestion
### 2. Update the operator-sdk-samples repo 
```

brew bump-formula-pr --strict --url=$OPERATORSDKURL --sha256=$OPERATORSUM operator-sdk
```

> **NOTE**  If there were any changes made to the CLI commands, make sure to look at the existing tests, in case they need updating.

Reviewer 2: It doesn't look like this was resolved. And actually for consistency with the rest of the doc's notes, I'd suggest this:

```suggestion
**Note:**  If there were any changes made to the CLI commands, make sure to look at the existing tests, in case they need updating.
```

    $ git push --tags
    ```

### 2. Updating the release notes

> Add the following line to the top of the GitHub release notes for `v1.3.0`:

Reviewer 2: Now that we're adding a line to the release notes in two places, we probably need to consider how we want those lines formatted after the second line is added.

As is, these instructions would result in the, which is inconsistent and repetitive:

```md
**Note:** See how to upgrade your project to the version v1.3.0 by checking the [Version Upgrade Guide](../migration/version-upgrade-guide.md#v1.3.0)
**NOTE:** ensure the `v1.3.0` tag is referenced when referring to sample code in the [SDK Operator samples repo](https://github.com/operator-framework/operator-sdk-samples/tree/v1.3.0) for this release. Links in SDK documentation are currently set to the samples repo `master` branch.
```

I looked at past releases' notes and we're pretty inconsistent about the formatting and location of the line about the Version Upgrade Guide. And I didn't see any lines about using the correct sample code tag. I think we only started tagging the sample repo somewhat recently, so that might explain that.

My point is, we don't have any working examples for how this _could_ or _should_ look when all is said and done. So I'm open to suggestions.

/cc @estroz

Author: IHMO:
- A user will check the info the specific release so will not be respective. 
- Regards he consistence: Could we not just:`**Note:** Ensure the ` for the samples one which will be less effort to change?

Author: Hi @joelanford and @estroz, 

The goal of this PR was just to add the steps that we agreed in the ENG meeting 
- Create the JIRA
- + nit regards the NOTEs in the releases 

Could we move forward with as it is and do further updates in others PR if required? OR what you think we should address here to get this merged?

Author: Hi @folks,

Should we move forward with it or not? Also, I think we did not create the tasks to upgrade the samples in the latest released which was the goal/problem that we are trying to address here. WDYT?

Reviewer: Adding migration guide entries is handled by the changelog/upgrade guide checker now. Instead of all these changes, just make a note about this and LGTM.

    $ git push --tags
    ```

> 

Reviewer: Why is this section being removed?

Author: It shows that we should add a reference of the samples in the operator-SDK release notes. Am I right? Have we not one it so far. Should we really do it?

Author: We did not it so far. I removed by mistake. 
Added back.

    $ git push --tags
    ```

> Then, add the following line to the top of the GitHub release notes for `v1.3.0`:

Reviewer 2: I didn't realize that we already have these post-release sections in the doc (I clearly didn't play close enough attention to this during the `v0.12.0` release :grimacing:)

But I think the existing sections capture the intent well enough already. The only changes that are necessary AFAICT, is that we should say that the actual post release steps by the releaser should be to create the tickets to do these, not actually do them.

WDYT?

Author: I updated it and keep it simple. Let's me know if you are OK with now.

```md

> **NOTE:** ensure the `v1.3.0` tag is referenced when referring to sample code in the [SDK Operator samples repo](https://github.com/operator-framework/operator-sdk-samples/tree/v1.3.0) for this release. Links in SDK documentation are currently set to the samples repo `master` branch.

Reviewer 2: Add another line that either:
1. Has a link to the section of the version upgrade guide that corresponds to this release.
2. Says that the release is fully backwards compatible and should require no migration steps.

Author: This part is speaking about the Release Notes of the samples repo and not over SDK. So, do you think that is really required we add the link for the migration doc on it as well? Should not make sense have it just in the SDK repo? 

Regards: `Says that the release is fully backwards compatible and should require no migration steps.`
A step always will be required which is pinned the SDK new version in the go.mod so IHMO it is not required.

Reviewer: This step refers to the SDK's release notes, since we don't have release notes for the samples repo. This should be its own post-release step. The title should be `Updating the Operator SDK release notes` to be explicit about which release notes need updating.

Author: Done ;-)

[homebrew-readme]:https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md#to-submit-a-version-upgrade-for-the-foo-formula
[homebrew-repo]:https://github.com/Homebrew/homebrew-core
[sdk-samples-repo]:https://github.com/operator-framework/operator-sdk-samples

> [upgrade-version-doc]:../migration/version-upgrade-guide.md

Reviewer 2: Nit
```suggestion
[version-upgrade-guide]:../migration/version-upgrade-guide.md
```

[ansible-guide]:../ansible/user-guide.md
[helm-guide]:../helm/user-guide.md
[ansible-memcached-sample]:https://github.com/operator-framework/operator-sdk-samples/tree/master/memcached-operator
[helm-memcached-sample]:https://github.com/operator-framework/operator-sdk-samples/tree/master/helm/memcached-operator

> [go-memcached-sample]:https://github.com/operator-framework/operator-sdk-samples/tree/master/memcached-operator

Reviewer 2: What about the other operator examples? If they're present, I think we should keep maintaining them. 

@hasbro17 @estroz @fabianvf Should we keep the other sample operators around?

Also, should we move the Go memcached-operator into a `go` subdirectory to match the other project type locations?

Author: Hi @joelanford, 

We have been unable to keep maintained the Memcached ones. IHMO we should not have many samples and we should have just the Memcached ones and then, we should actually improve them to be a very good sample by such as adding tests, coveralls, ci and etc . I added this topic for we discussed in the tech meeting.

Reviewer 2: I agree @camilamacedo86. Interested in what others think :wink:
